redis/src/main/scala/akka/streams/alpakka/redis/impl/RedisStreamingSourceStage.scala

```diff
@@ -0,0 +1,174 @@
+
+    private val maxConcurrency = settings.maxConcurrency
+
+    private var currentRequests = 0
+
+    private val maxBufferSize = settings.maxBufferSize
+
+    private val buffer: util.ArrayDeque[RedisKeyValue[K, V]] = new java.util.ArrayDeque[RedisKeyValue[K, V]]()
+
+    private val successCallback: AsyncCallback[RedisKeyValue[K, V]] =
+      getAsyncCallback[RedisKeyValue[K, V]](handleSuccess)
+
+    private val failureCallback: AsyncCallback[Exception] = getAsyncCallback[Exception](handleFailure)
+
+    private var redisFutureCompleted = false
+
+    val result: RedisFuture[lang.Long] = resultHandler(() => checkConcurrency(), successCallback, failureCallback)
+
+    result.thenAcceptAsync(new Consumer[lang.Long] {
+      override def accept(t: lang.Long): Unit =
+        redisFutureCompleted = true
+    })
+
+    setHandler(
+      out,
+      new OutHandler {
+
+        override def onDownstreamFinish(): Unit =
+          super.onDownstreamFinish()
+
+        override def onPull(): Unit =
+          if (!buffer.isEmpty) {
+            push(out, buffer.poll())
+          } else if (shouldTerminateStage) {
+            completeStage()
+          }
+      }
+    )
+
+    override def postStop(): Unit =
+      super.postStop()
+
+    def handleSuccess(result: RedisKeyValue[K, V]): Unit = {
+
+      currentRequests = currentRequests - 1
+
+      (buffer.isEmpty, isAvailable(out)) match {
+        case (false, true) =>
+          push(out, buffer.poll())
+          buffer.offer(result)
+        case (true, true) =>
+          push(out, result)
+        case (_, false) =>
+          buffer.offer(result)
+      }
+      if (shouldTerminateStage) {
+        completeStage()
+      }
+
+    }
+
+    private def shouldTerminateStage: Boolean =
+      redisFutureCompleted &&
+      currentRequests == 0 &&
+      buffer.isEmpty
+
+    def checkConcurrency(): Boolean = {
+      currentRequests = currentRequests + 1
+      maxBufferSize > buffer.size &&
+      maxConcurrency > currentRequests
+    }
+
+    def handleFailure(ex: Exception): Unit =
+      failStage(ex)
+
+  }
+
+  override def shape: SourceShape[RedisKeyValue[K, V]] = SourceShape.of(out)
+}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[redis] class RedishmgetSourceStage[K, V](key: K,
+                                                 fields: Seq[K],
+                                                 val connection: StatefulRedisConnection[K, V],
+                                                 val settings: RedisSubscriberSettings =
+                                                   RedisSubscriberSettings.Defaults)
+    extends RedisStreamingSourceStage[K, V] {
+
+  def resultHandler(checkConcurrency: () => Boolean,
+                    successCallback: AsyncCallback[RedisKeyValue[K, V]],
+                    failureCallback: AsyncCallback[Exception]): RedisFuture[lang.Long] =
+    connection
+      .async()
```

Comment:
I am not too familiar with lettuce API, but can we use `connection.reactive()` here and simplify this GraphStage?

Reply:
I have tried it out, and it is possible to adapt `hmget` to akka-stream `Source` directly using the reactive lettuce client:

```scala
val source: Source[KeyValue[K, V], NotUsed] =
  Source.fromPublisher(connection.reactive().hmget(key, fields: _*))
```